File: Shipwright/cmb3d/asset/zar.cpp
#include "zar.h"
#include <cstring>

namespace Zelda3D {

static uint16_t u16(const uint8_t* b, size_t o) { uint16_t v; memcpy(&v, b + o, 2); return v; }
static uint32_t u32(const uint8_t* b, size_t o) { uint32_t v; memcpy(&v, b + o, 4); return v; }
static std::string cstr(const uint8_t* b, size_t o, size_t cap) {
    std::string s;
    for (size_t i = o; i < cap && b[i]; i++) s.push_back((char)b[i]);
    return s;
}

Zar::Zar(std::vector<uint8_t> data) : mData(std::move(data)) {
    const uint8_t* b = mData.data();
    size_t n = mData.size();
    if (n < 0x18 || memcmp(b, "ZAR\x01", 4) != 0) { mErr = "not a ZAR archive"; return; }
    uint16_t nTypes = u16(b, 0x08);
    uint16_t nFiles = u16(b, 0x0A);
    uint32_t typesOff = u32(b, 0x0C);
    uint32_t metaOff = u32(b, 0x10);
    uint32_t dataOff = u32(b, 0x14);

    mFiles.resize(nFiles);
    for (uint16_t i = 0; i < nFiles; i++) {
        uint32_t off = u32(b, dataOff + 4 * i);   // per-file data offset
        uint32_t m = metaOff + 8 * i;
        uint32_t fsize = u32(b, m);
        uint32_t nameStr = u32(b, m + 4);
        mFiles[i].name = cstr(b, nameStr, n);
        mFiles[i].size = fsize;
        mFiles[i].offset = off;
    }
    for (uint16_t t = 0; t < nTypes; t++) {
        uint32_t e = typesOff + 16 * t;
        uint32_t cnt = u32(b, e);
        uint32_t idxOff = u32(b, e + 4);
        std::string tname = cstr(b, u32(b, e + 8), n);
        if (idxOff == 0xFFFFFFFF) continue;
        for (uint32_t k = 0; k < cnt; k++) {
            uint32_t fi = u32(b, idxOff + 4 * k);
            if (fi < mFiles.size()) mFiles[fi].type = tname;
        }
    }
    mOk = true;
}

const ZarFile* Zar::firstOfType(const std::string& type) const {
    for (const auto& f : mFiles)
        if (f.type == type) return &f;
    return nullptr;
}

const ZarFile* Zar::firstWithSuffix(const std::string& suffix) const {
    for (const auto& f : mFiles)
        if (f.name.size() >= suffix.size() &&
            f.name.compare(f.name.size() - suffix.size(), suffix.size(), suffix) == 0)
            return &f;
    return nullptr;
}

std::vector<uint8_t> Zar::read(const ZarFile& f) const {
    if ((size_t)f.offset + f.size > mData.size()) return {};
    return std::vector<uint8_t>(mData.begin() + f.offset, mData.begin() + f.offset + f.size);
}

} // namespace Zelda3D

File: tools/zar.py
#!/usr/bin/env python3
"""Parser for OoT3D ZAR archives (the container holding CMB models, textures, etc).

Format: CloudModding 3D:ZAR_format, verified against zelda_tsubo.zar.
Files inside a ZAR are NOT compressed in OoT3D (the on-disk RomFS .zar is plain);
each file's data is located via the per-file data-offset table in the data section.
"""
from __future__ import annotations
import struct
from dataclasses import dataclass


def _u16(b, o): return struct.unpack_from("<H", b, o)[0]
def _u32(b, o): return struct.unpack_from("<I", b, o)[0]
def _cstr(b, o):
    e = b.index(b"\x00", o)
    return b[o:e].decode("ascii", "replace")


@dataclass
class ZarFile:
    name: str
    type: str
    size: int
    offset: int   # offset into the ZAR blob


class Zar:
    def __init__(self, data: bytes):
        self.data = data
        if data[0:4] != b"ZAR\x01":
            raise ValueError("not a ZAR archive")
        self.size       = _u32(data, 0x04)
        n_types         = _u16(data, 0x08)
        n_files         = _u16(data, 0x0A)
        types_off       = _u32(data, 0x0C)
        meta_off        = _u32(data, 0x10)
        data_off        = _u32(data, 0x14)
        # per-file data offsets live at the start of the data section
        self._data_offsets = [_u32(data, data_off + 4 * i) for i in range(n_files)]
        # file metadata: size + name
        self.files: list[ZarFile] = []
        for i in range(n_files):
            m = meta_off + 8 * i
            fsize = _u32(data, m)
            fname = _cstr(data, _u32(data, m + 4))
            self.files.append(ZarFile(fname, "", fsize, self._data_offsets[i]))
        # file types: assign a type string to each file index
        for t in range(n_types):
            e = types_off + 16 * t
            cnt = _u32(data, e)
            idx_off = _u32(data, e + 4)
            tname = _cstr(data, _u32(data, e + 8))
            if idx_off == 0xFFFFFFFF:
                continue
            for k in range(cnt):
                fi = _u32(data, idx_off + 4 * k)
                self.files[fi].type = tname

    def read(self, f: ZarFile) -> bytes:
        return self.data[f.offset: f.offset + f.size]

    def by_type(self, t: str):
        return [f for f in self.files if f.type == t]


if __name__ == "__main__":
    import sys
    path = sys.argv[1] if len(sys.argv) > 1 else "scratch/extract/zelda_tsubo.zar"
    z = Zar(open(path, "rb").read())
    print(f"{path}: {len(z.files)} files, archive size {z.size}")
    for f in z.files:
        print(f"  [{f.type or '?':6}] {f.name:32} size={f.size:>8} @0x{f.offset:x}")
